macros.lisp

```diff
@@ -10,44 +10,134 @@ If TEST is `jtruep' evaluate THEN, otherwise evaluate ELSE."
        ,then
        ,else))
 
-(defmacro if_ (test then &optional (else nil))
-  "JSON-aware version of `cl:if'.
-If TEST is `jtruep' evaluate THEN, otherwise evaluate ELSE.
-Alias for `jif'."
-  (warn 'deprecated :deprecated 'if_ :replacement "NJSON/ALIASES:TRUE")
-  `(jif ,test ,then ,else))
-
 (defmacro jwhen (test &body body)
   "JSON-aware version of `cl:when'.
 If TEST is `jtruep' evaluate BODY."
   `(jif ,test
        (progn ,@body)))
 
-(defmacro when_ (test &body body)
-  "JSON-aware version of `cl:when'.
-If TEST is `jtruep' evaluate BODY.
-Alias for `jwhen'."
-  (warn 'deprecated :deprecated 'when_ :replacement "NJSON/ALIASES:when")
-  `(jwhen ,test ,@body))
-
 (defmacro jor (&rest args)
   "JSON-aware version of `cl:or'."
   `(or ,@(loop for arg in args
                collecting `(jwhen ,arg ,arg))))
 
-(defmacro or_ (&rest args)
-  "JSON-aware version of `cl:or'.
-Alias for `jor'."
-  (warn 'deprecated :deprecated 'or_ :replacement "NJSON/ALIASES:OR")
-  `(or ,@args))
-
 (defmacro jand (&rest args)
   "JSON-aware version of `cl:and'."
   `(and ,@(loop for arg in args
                 collecting `(jwhen ,arg ,arg))))
 
-(defmacro and_ (&rest args)
-  "JSON-aware version of `cl:and'.
-Alias for `jand'."
-  (warn 'deprecated :deprecated 'and_ :replacement "NJSON/ALIASES:AND")
-  `(and ,@args))
+(defun check-value (expected indices object)
+  "Check that JSON value in FORM at INDICES is `equal' to EXPECTED specification."
+  (let ((result (jget indices object t)))
+    (or (case expected
+          (:true (eq t result))
+          (:false (eq nil result))
+          (array (arrayp result))
+          (null (hash-table-p result))
+          (t (equal result expected)))
+        (error 'value-mismatch
+               :expected expected :actual result
+               :object (jget (subseq indices 0
+                                     ;; FIXME: Not exactly a perfect heuristic?
+                                     (max 0 (- (length indices) 2)))
```

Comment:
Why remove 2?  What heuristic is it?

Reply:
This is an attempt at getting the more meaningful parent object for EXPECTED by stripping off the last index. But yeah, two doesn't make much sense because the right end of `subseq` is exclusive already. I've switched to `1-` there.

Reply:
Please add a comment to explain this.